search: frontend logic for indexed structural search

rvantonder opened the pull request:
Adds the frontend logic to call structural search over the set of repos.

- Zoekt returns matches that satisfy constant strings in the structural search pattern
- From these matches, build a map of `repo => filtered file list`
- Only call searcher over those repos in the map and add the filtered file list, per repo, to the `IncludePatterns` for structural search

Test plan: Added a test to check that the structural search path filters (mocked) indexed files only.

Notes: the filtered file list could potentially be large, both for allocating and for sending over the wire. [We have options](https://docs.google.com/a/sourcegraph.com/document/d/16WwZA8zf-soI0HgCXGrRL5IrmyZnGYrB5VTXsqguUzo/edit?disco=AAAADln3Bgg) but I'm not addressing those in this PR, and would first like to get the end-to-end codepath.

unknwon commented on cmd/frontend/graphqlbackend/textsearch.go:
```diff
@@ -580,6 +589,11 @@ func searchFilesInRepos(ctx context.Context, args *search.Args) (res []*fileMatc
 
+				if args.Pattern.IsStructuralPat && searcherReposFilteredFiles != nil {
+					// Modify the search query to only run for the filtered files
+					args.Pattern.IncludePatterns = append(args.Pattern.IncludePatterns, searcherReposFilteredFiles[string(repoRev.Repo.Name)]...)
```
Is there a chance or what happens when `searcherReposFilteredFiles[string(repoRev.Repo.Name)]` is nil?

rvantonder replied:
It's not possible because `searcherRepos` and the name/keys for `searcherReposFilteredfiles` are kept in sync for `IsStructuralPat = true`. I'll add a (redundant for now) check in case any other callers of this function break the invariant.

unknwon commented on cmd/frontend/graphqlbackend/textsearch.go:
```diff
@@ -646,13 +659,50 @@ func searchFilesInRepos(ctx context.Context, args *search.Args) (res []*fileMatc
-		addMatches(matches)
+
+		if args.Pattern.IsStructuralPat {
+
```
```suggestion
```

keegancsmith commented:
Zoekt has an overall limit on results returned, as well as a per shard limit. So you may not have a map of everything. This is a good thing though, since we don't want to have unbounded memory growth. But if we were to apply this codepath to `replace` then we would want every possible match.

Our zoekt fork contains two extensions that may interest you:
- `type:file` will only return path file matches. IE `type:file foo` returns all files which has `foo` in contents.
- `zoekt.Searcher.List` supports taking a query, and then will return repos matching that query. ie `foo` will return all repos which have `foo` in a files contents in the repo.

keegancsmith commented on cmd/frontend/graphqlbackend/textsearch.go:
```diff
@@ -623,7 +637,6 @@ func searchFilesInRepos(ctx context.Context, args *search.Args) (res []*fileMatc
 		defer wg.Done()
 		matches, limitHit, reposLimitHit, err := zoektSearchHEAD(ctx, args, zoektRepos, false, time.Since)
 		mu.Lock()
-		defer mu.Unlock()
```
can you add back the defer here. It makes me a bit uneasy to have so much code that we need to remember to call Unlock on. And from what I can tell you want the rest of the function to be under the lock.

unknwon replied:
Inside `callSearcherOverRepos` has a call to `mu.Lock()` (line 603), wouldn't that deadlock?

keegancsmith commented on cmd/frontend/graphqlbackend/textsearch.go:
```diff
@@ -646,13 +659,50 @@ func searchFilesInRepos(ctx context.Context, args *search.Args) (res []*fileMatc
+			// structural search is concerned, so the lock can be
+			// freely released.
+			mu.Unlock()
+			searchErr = callSearcherOverRepos(repos, partition)
```
searchErr can only be written to with the lock.

keegancsmith commented on cmd/frontend/graphqlbackend/textsearch.go:
```diff
@@ -646,13 +659,50 @@ func searchFilesInRepos(ctx context.Context, args *search.Args) (res []*fileMatc
+			// The Zoekt part of the search is done here as far as
+			// structural search is concerned, so the lock can be
+			// freely released.
+			mu.Unlock()
```
you can call Unlock here, then call `callSearcherOverRepos` then call `Lock` again. That may be easier to understand than the current code.

Additionally no point holding the lock while we build the above map structures.

Is there not a way to restructure this code such that it is more like a pipeline, rather than adding in this sort of extra logic?

rvantonder replied:
Addressed, thanks. See main comment chain for restructuring things.

rvantonder commented:
> Is there not a way to restructure this code such that it is more like a pipeline, rather than adding in this sort of extra logic?

I want this as much as you do. I don't see a way right now.  Just with #6197, which was supposed to be semantics preserving, resulted in a bunch of issues and took much longer than I would have liked it to. In summary, for such a restructure:

- The `callSearcherOverRepos` really does need to be pulled out to create such a pipeline, but with all the state over closures/goroutines in this function, pulling this out entirely seems trickier compared to #6197 (I still think it is). The other idea was a duplication of `searchFilesInRepos` for structural search, that strips out some of the logic here. 
- I don't think duplicating the function solves the problem overall, because much of the logic for structural search _does_ share the concerns in this function (e.g., in the near future, we would want to fan out to searcher for unindexed search as well--the guard around that will be removed, and things should 'just work').
- To solve, we should make the processes separable here (call over searcher, call over zoekt), and that means teasing out the shared state/utility functions and locking operations.

I agree we should attack this problem, but it will need to be a dedicated effort and the complexity seems too open-ended to attempt a restructure now. Let's revisit and plan out this effort when we're both free again.

rvantonder commented:
This component is developed alongside a larger diff #6387, which I split for review. I realized that the approved state of this diff has the problem that:
- modifying `includePatterns` of the original args when the structural search flag is _on_ is racy because we fan goroutines to searcher inside the loop. 

This affects only structural search.

The recent commit a234991 fixes this to create a deepcopy of the `includePatterns` when structural search is active. I'm including it in this PR rather than another because it is closely related.

@slimsag advised on the code for the deepcopy. @slimsag could you just check the commit a234991 covers the desired behavior? I split up the variable declaration and copy operation so that the copy only happens when structural search is on, there is no need to do so otherwise.

rvantonder commented on cmd/frontend/graphqlbackend/textsearch.go:
```diff
@@ -584,18 +587,21 @@ func searchFilesInRepos(ctx context.Context, args *search.Args) (res []*fileMatc
+				// Modify the search query to only run for the filtered files
+				if v, ok := searcherReposFilteredFiles[string(repoRev.Repo.Name)]; ok {
+					args.Pattern = &patternCopy
+					includePatternsCopy := make([]string, len(args.Pattern.IncludePatterns))
```
In fact, I think the lexical scope of `includePatternsCopy` should be pulled out and initialized at the for-loop iteration. Updated.

unknwon commented on cmd/frontend/graphqlbackend/textsearch.go:
```diff
@@ -561,6 +570,10 @@ func searchFilesInRepos(ctx context.Context, args *search.Args) (res []*fileMatc
 		for _, repoRev := range searcherRepos {
+			args := *args
+			patternCopy := *args.Pattern
+			var includePatternsCopy []string
```
It seems these variables are not used until line 591, what about move them to there? This way we can keep define & use close.

rvantonder replied:
Ack, makes sense.

unknwon commented on cmd/frontend/graphqlbackend/textsearch.go:
```diff
@@ -575,6 +588,16 @@ func searchFilesInRepos(ctx context.Context, args *search.Args) (res []*fileMatc
+			if args.Pattern.IsStructuralPat && searcherReposFilteredFiles != nil {
+				// Modify the search query to only run for the filtered files
+				if v, ok := searcherReposFilteredFiles[string(repoRev.Repo.Name)]; ok {
+					args.Pattern = &patternCopy
```
It seems what's been done with the use of `patternCopy` is saying:

```go
(*args).Pattern = (*args).Pattern
```

I can't tell if this is actually going to work without running the code.

Maybe try to print pointer address with `fmt.Printf("%T\n", <T>)` to help verify?

rvantonder replied:
Yes, where `*args` is copied. It works as expected when running (and @slimsag can take a glance to confirm).

unknwon replied:
Thanks for the confirm!

rvantonder commented:
> Another idea about searchErr race, what about using atomic.Value, it looks like a perfect fit in this case.

There's no `searchErr` race, that's addressed. I didn't change anything about this since, so, for now, let's not continue on that thread, we can revisit.

unknwon commented on cmd/frontend/graphqlbackend/textsearch.go:
```diff
@@ -575,6 +584,19 @@ func searchFilesInRepos(ctx context.Context, args *search.Args) (res []*fileMatc
+					copy(includePatternsCopy, args.Pattern.IncludePatterns)
+					args.Pattern.IncludePatterns = append(includePatternsCopy, v...)
+				}
+			}
```
Regarding `patternCopy` and `includePatternsCopy`, could we narrow down their scopes if they're only used as temp vars? For example:

```diff
			args := *args
-			patternCopy := *args.Pattern
-			var includePatternsCopy []string
			if args.Pattern.IsStructuralPat && searcherReposFilteredFiles != nil {
				// Modify the search query to only run for the filtered files
				if v, ok := searcherReposFilteredFiles[string(repoRev.Repo.Name)]; ok {
+					patternCopy := *args.Pattern
					args.Pattern = &patternCopy
+					includePatternsCopy := make([]string, len(args.Pattern.IncludePatterns))
					copy(includePatternsCopy, args.Pattern.IncludePatterns)
					args.Pattern.IncludePatterns = append(includePatternsCopy, v...)
				}
			}
```

WDYT?

rvantonder replied:
👍 Yeah this looks OK

rvantonder replied:
Thanks for taking another look :)

![](https://media0.giphy.com/media/3oz8xIsloV7zOmt81G/giphy.gif?cid=5a38a5a2f04ce02531a7dab8866bd60aceb418e229e2c2bc&rid=giphy.gif)

unknwon replied:
![](https://media1.giphy.com/media/bz7Lkscqm6mrK/giphy.gif?cid=5a38a5a250815eeac3eeff494ef346f7ca7b0a7b39c8fb00&rid=giphy.gif)

unknwon commented:
> There's no `searchErr` race, that's addressed. I didn't change anything about this since, so, for now, let's not continue on that thread, we can revisit.

Yeah, I know we managed to use mutex to avoid race, just seems tedious with `Lock`/`Unlock` calls we have. Agreed we should revisit at later time.
